[ethanwangkangen] iP

ethanwangkangen opened the pull request:
# Echo - your personal task manager!

**Features:**

- Adding and removing tasks
- Marking and unmarking tasks (as completed)
- Task types
- Save tasks on hard drive

**Task Types:**
1. Todo
2. Deadline: able to specify deadline date 
     (processed as `LocalDateTime` objects)
3. Event: able to specify start and end times

**Run the Echo chatbot by running main**
``` java 
public static void main (String args[]) throws InvalidCommandError, EmptyDescriptionError {
        Echo echo = new Echo();
        echo.run();
    }
``` 

**Project milestones:**
- [x] Tasks types and inputting dates
- [ ] Graphical User Interface: Coming soon


> [!NOTE]
> Currently has no GUI, only text commands allowed in terminal.
> Saves tasks to a file called tasks.ser

[Click me](https://www.youtube.com/watch?v=dQw4w9WgXcQ&skip_registered_account_check=true):grinning:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ethanwangkangen committed:
Create basic chatbot skeleton

ethanwangkangen committed:
Add echoing function to chatbot

ethanwangkangen committed:
Add list function to chatbot

ethanwangkangen committed:
Add mark as done

ethanwangkangen committed:
Add Todo, Deadline, Event

ethanwangkangen committed:
Add exceptions

ethanwangkangen committed:
Add delete function

ethanwangkangen committed:
Add testing script

ethanwangkangen committed:
Add save function

ethanwangkangen committed:
Add javadocs

ethanwangkangen committed:
Merge branch-Level-7 into master

ethanwangkangen committed:
Add dateTime feature

ethanwangkangen committed:
Merge branch-Level-8

ethanwangkangen committed:
Refactor into more classes

ethanwangkangen committed:
Add into echochat package

ethanwangkangen committed:
Merge branch 'master' into add-gradle-support

ethanwangkangen committed:
Fix issue with gradle

ethanwangkangen committed:
Add Junit tests

ethanwangkangen committed:
Add JavaDocs

ethanwangkangen committed:
Reformat code

ethanwangkangen committed:
Change code to fit coding standard

ethanwangkangen committed:
Add find command

ethanwangkangen committed:
Change code to fit coding standard again

ethanwangkangen committed:
Merge branch 'branch-A-CodingStandard'

ethanwangkangen committed:
Merge JavaDoc branch

ethanwangkangen committed:
Add getDesc() function to Command

ethanwangkangen committed:
Finish Level 9

ethanwangkangen committed:
Merge branch level 9

tansanchian commented on src/main/java/echochat/ChatBot.java:
```diff
@@ -0,0 +1,73 @@
+
+                switch (command.getType()) {
+                case EXIT:
+                    ui.goodbye();
```
Perhaps it would be better if a verb is used here?

tansanchian commented on src/main/java/echochat/ChatBot.java:
```diff
@@ -0,0 +1,73 @@
+                    continue;
+                }
+
+                switch (command.getType()) {
```
Would it be clearer if the method is named as getCommandType?

tansanchian commented on src/main/java/echochat/ChatBot.java:
```diff
@@ -0,0 +1,73 @@
+                    ui.showTaskList(taskList.getTaskList());
+                    break;
+                case MARK:
+                    Task taskToMark = taskList.getTaskList().get(command.getIndex() - 1);
```
Maybe getTask would be better than task to not confuse with getIndex?

tansanchian commented on src/main/java/echochat/ChatBot.java:
```diff
@@ -0,0 +1,73 @@
+                    ui.showTaskDeleted(removedTask, taskList.getTaskList().size());
+                    break;
+                case FIND:
+                    ui.showFoundTasks(command.getDescription(), taskList.getTaskList());
```
Maybe showTasksFound would be a more sensible order?

tansanchian commented on src/main/java/echochat/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+     * @return Task description with date included.
+     */
+    @Override
+    public String getDesc() {
```
I'm not sure what Desc meant here, maybe consider changing to a more straightforward name?

tansanchian reviewed (commented):
It's great overall! Just a few nits to fix.

SherChew27 commented on src/main/java/echochat/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Deadline extends Task{
+    protected String by;
+    protected LocalDateTime dt;
```
can be renamed to dateTime to aid in readability

SherChew27 commented on src/main/java/echochat/ChatBot.java:
```diff
@@ -0,0 +1,73 @@
+                    taskList.addToList(newTask);
+                    ui.showTaskAdded(newTask, taskList.getTaskList().size());
+                    break;
+                case UNKNOWN:
```
do you still need an unknown command if your default already handles that case?

SherChew27 commented on src/main/java/echochat/Storage.java:
```diff
@@ -0,0 +1,44 @@
+        try (FileInputStream fileIn = new FileInputStream(FILE_NAME);
+             ObjectInputStream in = new ObjectInputStream(fileIn)) {
+            lst = (ArrayList<Task>) in.readObject();
+        } catch (Exception e) {
```
is it possible for other exceptions to be thrown? if so, will it be handled?

SherChew27 commented on src/main/java/echochat/Task.java:
```diff
@@ -0,0 +1,83 @@
+import java.time.format.DateTimeParseException;
+
+public class Task implements Serializable {
+    private Boolean done;
```
this field is a boolean, and so should be renamed to isDone

SherChew27 commented on src/main/java/echochat/TaskList.java:
```diff
@@ -0,0 +1,46 @@
+
+public class TaskList {
+
+    private ArrayList<Task> lst = new ArrayList<>();
```
this field can be renamed to taskList instead of just lst so it is easier to understand

SherChew27 reviewed (commented):
Overall well done! Just some minor changes needed such as the naming of fields

ethanwangkangen committed:
Add GUI to chatbot

ethanwangkangen committed:
Merge branch-Level-10 into master

ethanwangkangen committed:
Add assertions to Parser class for internal consistency checks

- Added assertions to ensure that input parsing logic and task creation follow expected conditions. These checks validate assumptions that should always hold true during the execution of the program.

  - Added assertion after splitting the user input to ensure that the `parts` array contains at least one element.
  - Added assertion to verify that the `type` parameter passed to `parseTask` is either "todo", "deadline", or "event", ensuring that only valid task types are processed.
  - Added assertion to ensure that a non-null task is created before returning from the `parseTask` method.
  - For `Deadline` tasks, added an assertion to check that the "by" date is not null.
  - For `Event` tasks, added assertions to check that both "from" and "to" dates are present.

- These assertions will help catch logical errors and invalid internal states during development and debugging.
- Assertions are meant for development/debugging environments and will not be triggered in production unless explicitly enabled.

This change ensures that invalid inputs or internal inconsistencies are caught early during the development phase, leading to more robust and reliable code.

ethanwangkangen committed:
Improve code quality

ethanwangkangen committed:
Parser class: add assertions for internal consistency checks

The parsing logic in the Parser class assumes certain conditions
(e.g. valid task types, non-null task creation) but does not explicitly
enforce them.

Failing to enforce these assumptions could lead to silent failures or
unexpected behavior when the input is not as expected, particularly in
edge cases.

Let's add assertions to enforce key assumptions in the input parsing
process:
- Assert that the `parts` array contains at least one element after
  splitting user input.
- Assert that the `type` parameter passed to `parseTask` is either
  "todo", "deadline", or "event".
- Ensure a non-null task is created before returning from `parseTask`.
- For `Deadline` tasks, ensure that the "by" date is not null.
- For `Event` tasks, ensure that both "from" and "to" dates are present.

Assertions are appropriate here as they allow us to verify that internal
invariants hold, helping to catch bugs early without affecting
production behavior.

ethanwangkangen committed:
Refactor Main and Parser Classes for Improved Maintainability

The Main class handles user interface setup and event handling, while the Parser class is responsible for parsing commands and creating tasks.

The Main class's start method mixes UI setup, formatting, and input handling, leading to a complex and difficult-to-maintain method. Similarly, the Parser class’s parse method combines parsing logic with task creation, which complicates testing and debugging.

Refactoring is needed to improve code readability and adherence to the Single Responsibility Principle (SRP). By separating these concerns, we ensure that each class and method has a clear, single purpose.

Let's:

    Extract setupStage method from Main to handle stage setup separately.
    Extract formatWindow method from Main to handle window formatting separately.
    Add assertions in Parser to ensure consistency:
        Verify that parsed input splits correctly.
        Ensure only valid task types are processed.
        Confirm that non-null tasks are created before returning.
        For Deadline tasks, check that the "by" date is not null.
        For Event tasks, ensure both "from" and "to" dates are provided.

This refactoring enhances maintainability by isolating responsibilities and ensuring consistency in the Parser class. It aligns with best practices for cleaner, more manageable code.

ethanwangkangen committed:
Merge pull request #2 from ethanwangkangen/branch-A-Assertions

Branch A-assertions

ethanwangkangen committed:
Merge branch 'master' into branch-A-CodeQuality

ethanwangkangen committed:
Merge pull request #3 from ethanwangkangen/branch-A-CodeQuality

Branch A-code quality

ethanwangkangen committed:
Add isDuplicate Method to ChatBot Class

The ChatBot class does not have a mechanism to prevent duplicate tasks from being added to the task list. This lack of functionality allows for multiple instances of identical tasks, which can clutter the task list and hinder user experience.

To improve task management and maintain a clean task list, we need to implement a check for duplicate tasks.

Add the isDuplicate method to the ChatBot class to identify and handle duplicate tasks before they are added. This method compares the new task with existing tasks based on their description and other properties.

Implement this method in the getResponse method to ensure that duplicate tasks are not added. If a task is detected as a duplicate, the user receives a notification to inform them of the duplicate entry.

This change enhances task management by preventing duplicate entries, which improves data integrity and overall user experience. By checking for duplicates before adding new tasks, the task list remains organized and user interactions become more efficient.

ethanwangkangen committed:
Merge branch 'master' of https://github.com/ethanwangkangen/ip

ethanwangkangen committed:
Update README

ethanwangkangen committed:
Add comments for extension

ethanwangkangen committed:
Change parser logic

ethanwangkangen committed:
Add Ui.png

ethanwangkangen committed:
Update README.md

ethanwangkangen committed:
Update README.md

ethanwangkangen committed:
Make final changes

ethanwangkangen committed:
Add echo.jar and fix bug
